Add inverse hessian approximation to the returned state

aymgal opened the pull request:
The `state` returned after running `ScipyMinimize` now contains the approximated inverse Hessian matrix (if the optimizer supports it, e.g. `'BFGS'`). The inverse Hessian matrix gives an approximation of the covariance matrix.
A small test has been added to check against original scipy. 
The goal is to solve issue #381.

mblondel commented on jaxopt/_src/scipy_wrappers.py:
```diff
@@ -312,10 +313,19 @@ def scipy_fun(x_onp: onp.ndarray) -> Tuple[onp.ndarray, onp.ndarray]:
+        if isinstance(res.hess_inv, onp.ndarray):
+            hess_inv = res.hess_inv
+        else:
+            hess_inv = res.hess_inv.todense()
```
Silently densifying sparse matrices is not ideal. This could blow up memory... It may be better to raise an exception for the time being.

mblondel replied:
JAX has an experimental sparse module https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html

aymgal replied:
Indeed, I understand why it's a problem to always densify it.
However, most of the time (at least in the test routines), the returned object is not a numpy array, but a `LbfgsInvHessProduct` instance (see [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.LbfgsInvHessProduct.html))... and I could not find a way to extract the matrix as sparse matrix.
Maybe just return the `LbfgsInvHessProduct` as it is? There is no JAX version of it, if I'm correct. I tried to get the hessian out of the operator without turning it into a dense matrix first (e.g. via a product with JAX's sparse matrix, but it returns a numpy dense array). 
Do you have a specific idea of how to use the sparse module from JAX in this context?

mblondel replied:
Sorry for the late reply. From what you said, simply adding the `LbfgsInvHessProduct` object (without converting it to an array) to the state seems like the way to go. What do you think?

aymgal replied:
@mblondel The problem is that this object cannot be cast by default to a JAX object. If I directly return the object itself, I get this error after running the optimizer `'L-BFGS-B'`:
```
TypeError: Value <2x2 LbfgsInvHessProduct with dtype=float64> with type <class 'scipy.optimize.lbfgsb.LbfgsInvHessProduct'> is not a valid JAX type
```
There should be a way to extract the matrix while keeping it as sparse in memory, but I can't figure out how. Or to wrap the scipy's `LbfgsInvHessProduct` object to a JAX-friendly one, but I am not sure how to do this either.

For `'BFGS'` there is no such problem, since the returned hessian is an actual matrix.

aymgal replied:
Or what about allowing the user to decide if the hess_inv should be returned or not? By default it would not be, but if the user knows that memory won't be a problem or if the number of parameters is low, it densifies the matrix and returns it.

mblondel replied:
I feel it should be possible to wrap `LbfgsInvHessProduct` into a pytree (see `RegisteredSpecial2` example [here](https://jax.readthedocs.io/en/latest/pytrees.html))

@froystig Any recommendation on this?

froystig replied:
Resources on custom pytrees registration:
* https://jax.readthedocs.io/en/latest/pytrees.html
* https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html

mblondel reviewed (commented):
LGTM apart from the remark below

aymgal commented:
Thanks @mblondel and @froystig for the tips. I added a new class to register the scipy's LbfgsInvHessProduct class as a pytree. It seems to work now, I also updated the test script. Let me know what you think!

mblondel commented on tests/scipy_wrappers_test.py:
```diff
@@ -601,6 +601,46 @@ def wrapper(box_len):
+    r_min, r_max = -5.0, 5.0
+    onp.random.seed(6574)
+    self.x0 = r_min + onp.random.rand(2) * (r_max - r_min)
+    #self.x0 = [1., 1.]
```
Remove

mblondel commented on tests/scipy_wrappers_test.py:
```diff
@@ -601,6 +601,46 @@ def wrapper(box_len):
+    self.assertAllClose(res_scipy.hess_inv, res_jaxopt.state.hess_inv)
+    # onp.testing.assert_almost_equal(res_scipy.hess_inv, 
+    #                                 res_jaxopt.state.hess_inv, 
+    #                                 decimal=6)
```
Remove commented assertion

mblondel reviewed (commented):
Apologies for the late reply. LGTM apart from some left-over comments.

By the way, could you squash the commits? Thanks!

aymgal commented:
Thanks for the feedback, I removed the commented lines and squashed the 5 commits.

mblondel commented:
There is a failure on line 51 `AttributeError: module 'scipy' has no attribute 'optimize'`. Not sure what's going on, this line should work.

aymgal commented:
@mblondel That was strange indeed, it was working on my machine. I changed the way in import `LbfgsInvHessProduct`, let's see if it works.
I also noticed that other tests that I missed before were failing, in relation to the LBFGS optimizer. Now it should work, hopefully.

aymgal commented:
@mblondel a test is still failing, but it does not seem to be related in any way to this PR. The failed test is `bisection_test.py`, which never relates to the `scipy_wrappers` module on which I've worked on.

froystig commented:
I believe that failure (indicating a jax tracer leak) is one that is currently known, also on the `main` branch.

froystig commented:
> I believe that failure (indicating a jax tracer leak) is one that is currently known, also on the `main` branch.

#412 ought to fix that failure

mblondel commented:
If you merge main in your branch, this should now work. Thank you for your patience.
